dot/telemetry/mailer.go

```diff
@@ -0,0 +1,132 @@
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/lib/genesis"
+	"github.com/gorilla/websocket"
+)
+
+const (
+	defaultMessageTimeout = time.Second
+	defaultMaxRetries     = 5
+	defaultRetryDelay     = time.Second * 15
+)
+
+var messageReceiver chan Message = make(chan Message, 256)
+
+type telemetryConnection struct {
+	wsconn    *websocket.Conn
+	verbosity int
+	sync.Mutex
+}
+
+// Handler struct for holding telemetry related things
+type mailer struct {
+	msg         chan Message
+	connections []*telemetryConnection
+	log         log.LeveledLogger
+}
+
+func newMailer() *mailer {
+	return &mailer{
+		msg: messageReceiver,
+		log: log.NewFromGlobal(log.AddContext("pkg", "telemetry")),
+	}
+}
+
+// BootstrapMailer setup the mailer, the connections and start the async message shipment
+func BootstrapMailer(ctx context.Context, conns []*genesis.TelemetryEndpoint) {
+	mlr := newMailer()
+
+	for _, v := range conns {
+		for connAttempts := 0; connAttempts < defaultMaxRetries; connAttempts++ {
+			c, _, err := websocket.DefaultDialer.Dial(v.Endpoint, nil)
+			if err != nil {
+				mlr.log.Debugf("issue adding telemetry connection: %s", err)
+				time.Sleep(defaultRetryDelay)
+				continue
+			}
+
+			mlr.connections = append(mlr.connections, &telemetryConnection{
+				wsconn:    c,
+				verbosity: v.Verbosity,
+			})
+			break
+		}
+	}
+
+	go mlr.asyncShipment(ctx)
```

Comment:
We should pass a ready channel which can be closed inside this goroutine, to signal to the caller of `BootstrapMailer` we're ready to process

Reply:
I don't think we need to pass a ready function because `asyncShipment` does not have any conditions to start process the channel,  and the channel is already created by calling `BootstrapMailer` so I think we must change the `BootstrapMailer` to return an error that means: _you told us to start the telemetry but we are not able to bootstrap the async shipment then the `SendMessage` function will put your message into the channel however we won't send your message_